### PR TITLE
Add lms to RustCrypto/signatures

### PR DESCRIPTION
**Pull request by incertia**
This adds the `lms` directory to RustCrypto/signatures under the crate name `lms-signature`. The authors of the [original RFC](https://datatracker.ietf.org/doc/html/rfc8554#section-5.4) seem quite happy using the phrase so this keeps the naming in the signatures repository consistent. Hopefully I ported the Github workflow without any issues.

There are two notable limitations to our implementation:

1. We rely on `generic_array::GenericArray` from the older `digest-0.10.7` instead of using the `hybrid_array::Array` approach from the upcoming `digest-0.11` ecosystem.
2. We lack a way of compressing Merkle tree data so we just store the entire tree which causes H15, H20, and H25 variants of LMS to not work due to storage limitations.

This PR also does not modify the main README in any way. I have also refrained from committing any changes to Cargo.lock.

### CODE REVIEW

File: .github/workflows/lms.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    paths:
+      - ".github/workflows/lms.yml"
+      - "lms/**"
+      - "Cargo.*"
+  push:
+    branches: master
+
+defaults:
+  run:
+    working-directory: .
```
Comment (tarcieri) [inline .github/workflows/lms.yml:13]: I would expect this to be:

```suggestion
    working-directory: lms
```

File: lms/.github/workflows/lms-rust.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+name: lms-rust
```
Comment (tarcieri) [inline lms/.github/workflows/lms-rust.yml:1]: Is this left over? It seems redundant with `.github/workflows/lms.yml`

File: lms/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+[package]
+name = "lms-signature"
+version = "0.1.0"
+edition = "2021"
```
Comment (tarcieri) [inline lms/Cargo.toml:4]:
```suggestion
edition = "2021"
license = "Apache-2.0 OR MIT"
repository = "https://github.com/RustCrypto/signatures/tree/master/lms"
readme = "README.md"
rust-version = "1.73"
categories = ["cryptography"]
keywords = ["crypto", "signature"]
```

File: lms/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[lib]
+name = "lms"
+path = "lms/lib.rs"
```
Comment (tarcieri) [inline lms/Cargo.toml:10]: Hmm, this is a lot of directories named `lms`, i.e. `lms/lms/lms`. Perhaps use a standard `src` directory instead?

File: lms/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+edition = "2021"
+
+[dependencies]
+anyhow = "1.0.75"
```
Comment (tarcieri) [inline lms/Cargo.toml:14]: `anyhow` is typically used for application-level error handling as opposed to libraries

File: lms/lms/lib.rs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+//! LMS in Rust
```
Comment (tarcieri) [inline lms/lms/lib.rs:1]: FWIW, pretty much all of our crates support `no_std`, though I'm happy to address that as a followup

File: lms/lms/lms/private.rs [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+use std::ops::Add;
+use typenum::{Sum, U28};
+
+/// Opaque struct representing a LMS public key
+///
+/// Note: there is no requirement to map specific LMS algorithms to specific
+/// LM-OTS algorithms so it must be parametrized. With the algorithms provided
+/// by this crate, this is done via
+/// [LmsSha256M32H10](crate::lms::LmsSha256M32H10)<[LmsOtsSha256N32W4](crate::ots::LmsOtsSha256N32W4)>.
+pub struct PrivateKey<Mode: LmsMode> {
```
Comment (tarcieri) [inline lms/lms/lms/private.rs:23]: As a recommendation for cross-crate consistency, we generally use the names `SigningKey` and `VerifyingKey`, which are named after their roles / what they do, a convention also adopted in `ed25519-dalek`

File: lms/Cargo.toml (hunk 5 above)
Comment (tarcieri) [reply lms/Cargo.toml:14]: It also looks like it only has one usage here: https://github.com/RustCrypto/signatures/pull/801/files#diff-4a60ec7cc2aeac6ce05a26ae2a5b3d69d23c199b6d4f9a93f2a2bef5ef90ba88R106

File: lms/lms/lib.rs (hunk 6 above)
Comment (incertia) [reply lms/lms/lib.rs]: I think our code is `no_std` compatible but it has not been explicitly verified. We can probably address this some time later.

File: lms/lms/lms/private.rs (hunk 7 above)
Comment (incertia) [reply lms/lms/lms/private.rs]: I have updated the names to `SigningKey` and `VerifyingKey`

File: lms/.gitignore [hunk 8]
```diff
@@ -0,0 +1,2 @@
+Cargo.lock
+target/
```
Comment (tarcieri) [inline lms/.gitignore:2]: This whole file can be removed as the toplevel `.gitignore` should address this (and also we check in `Cargo.lock` to have a deterministic build, which is the new recommendation from the Cargo team)

File: lms/CHANGELOG.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# Changelog
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## 0.1.0 (2024-01-30)
+- Initial release
```
Comment (tarcieri) [inline lms/CHANGELOG.md:8]: Leave this blank for now (the date is already invalid) and we can bump it with an actual release:

```suggestion
```

File: lms/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,38 @@
+[package]
+name = "lms-signature"
+version = "0.1.0"
```
Comment (tarcieri) [inline lms/Cargo.toml:3]: We tend to bump version in a release PR/commit, so for now I'd suggest:

```suggestion
version = "0.0.0"
```

File: lms/src/lms/private.rs [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+
+        let lms_priv = SigningKey::<LmsSha256M32H5<LmsOtsSha256N32W8>>::new_from_seed(id, seed);
+        let lms_pub = lms_priv.public();
+        assert_eq!(lms_pub.k(), expected_k);
+        assert_eq!(lms_pub.id(), &id);
+    }
+
+    #[test]
+    fn test_kat_2() {
+        let expected_signature = [
```
Comment (tarcieri) [inline lms/src/lms/private.rs:235]: Can you provide some information on the provenance of these KATs?

It appears to be [RFC8554 Appendix F](https://datatracker.ietf.org/doc/html/rfc8554#appendix-F) but I'm noticing `test_pk_tree_kat1` appears to be "Test Case 2".

File: lms/src/lms/public.rs [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+        }
+        if self.k == tmp {
+            Ok(())
+        } else {
+            Err(Error::new())
+        }
+    }
+}
+
+/// Converts a [VerifyingKey] into its byte representation
```
Comment (tarcieri) [inline lms/src/lms/public.rs:100]: To properly linkify in rustdoc, `VerifyingKey` needs to be surrounded with backticks:

```suggestion
/// Converts a [`VerifyingKey`] into its byte representation
```

File: lms/src/lms/signature.rs [hunk 13]
```diff
@@ -0,0 +1,318 @@
+//! Contains the [Signature] type
```
Comment (tarcieri) [inline lms/src/lms/signature.rs:1]: Similar concern to [this one](https://github.com/RustCrypto/signatures/pull/801/files#r1480048307) regarding backticks (and elsewhere in this file):

```suggestion
//! Contains the [`Signature`] type
```

File: lms/src/lms/private.rs (hunk 11 above)
Comment (incertia) [reply lms/src/lms/private.rs:235]: @tjade273

**Comment by tjade273** [comment]
@tarcieri do you have opinions on large in-file KAT buffers like [this](https://github.com/trailofbits/signatures/blob/8e74c6b877949b27583375ee7619ba2d94159d08/lms/src/lms/private.rs#L234)? 

Debating whether we should
1. Leave as-is
2. Break large KATs out to a separate directory
3. Just include a hash of the actual expected result and compare against that

There will be a couple more large KATs eventually once HSS is implemented, and if we want to support the larger merkle tree sizes then they will start getting really big.

**Comment by tarcieri** [comment]
@tjade273 we generally just check them in, but for very large ones omit them from the resulting `.crate`, e.g.:

https://github.com/RustCrypto/MACs/blob/43cc597/cmac/Cargo.toml#L14

**Review by tarcieri — APPROVED** [review]

File: .github/workflows/lms.yml [hunk 14]
```diff
@@ -0,0 +1,52 @@
+name: lms-rust
```
Comment (tarcieri) [inline .github/workflows/lms.yml:1]: Noticed this after I approved, would probably just suggest this match the crate name:

```suggestion
name: lms-signature
```

You might also consider naming this file the same as the crate name, i.e. `lms-signature.yml`

Comment (tarcieri) [reply .github/workflows/lms.yml:1]: Then I remembered we omit it with `ed448`. How about just calling it `lms`?

**Comment by tarcieri** [comment]
utACK, let me know if you'd like to make any additional changes or otherwise we can get this merged

**Comment by incertia** [comment]
We can probably get this merged.

**Review by tarcieri — APPROVED** [review]

**Pull request merged**